### PR TITLE
feat(Cypress): give basic structure to cypress typescript files

### PR DESCRIPTION
**Pull request by Sembauke**
This adds some basic structure to the Cypress TypeScript files. It gives a global `Selector` object which will be sub divided in 4 different segments (id, tag, data, class). Data as the defined `data-cy` as described in the Cypress documentation. This will make it easier to differentiate when writing new tests.

I think this is the best approach. 

I have changed this because TypeScript does not like having the same variable defined in different files. 

I've also converted the commands to TypeScript. The declarations had to be defined in a different file (`index.d.ts`).

e.g. : https://docs.cypress.io/guides/tooling/typescript-support#Types-for-Custom-Commands

Let me know if you have any feedback on the selectors.
Cheers.

### CODE REVIEW

File: cypress/support/types.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  class: {
+    outputText: '.output-text',
+    reactMonacoEditor: '.react-monaco-editor-container'
+  },
+  id: {},
+  data: {},
+  tag: {
+    inputSolution: 'input[name="solution"]'
+  }
+};
```
Comment (ojeytonwilliams) [inline cypress/support/types.ts:11]: This would make more sense if it was called 'selectors', since it doesn't contain any types.

**Review by raisedadead — APPROVED** [review]

File: cypress/support/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+     * @example cy.dataCy('greeting')
+     */
+    login(): Chainable<Element>;
+    preserveSession(): Chainable<Element>;
+    setPrivacyTogglesToPublic(): Chainable<Element>;
+    goToSettings(): Chainable<Element>;
+    typeUsername(username: string): Chainable<Element>;
+    resetUsername(): Chainable<Element>;
+  }
+}
```
Comment (ojeytonwilliams) [inline cypress/support/index.d.ts:16]: I'm not sure why the Cypress docs recommend this approach, but there's no need for you to figure out the types - you can infer them.

For example, you can put the declaration inside `commands.ts` and write things like this

```ts
const login = () => {
  cy.visit(`${Cypress.env('API_LOCATION')}/signin`);
  cy.contains('Welcome back');
}

Cypress.Commands.add('login', login);

declare global {
  namespace Cypress {
    interface Chainable {
      login: typeof login;
    }
  }
}
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: cypress/support/types.ts (hunk 1 above)
Comment (ojeytonwilliams) [reply cypress/support/types.ts:11]: Oops, I meant the entire file. It should be called `selectors.ts`, imo.

File: cypress/support/index.d.ts (hunk 2 above)
Comment (ojeytonwilliams) [reply cypress/support/index.d.ts:16]: So, this file can go.

File: cypress/support/commands.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+Cypress.Commands.add('preserveSession', preserveSession);
+
+Cypress.Commands.add('setPrivacyTogglesToPublic', setPrivacyTogglesToPublic);
+
+Cypress.Commands.add('goToSettings', goToSettings);
+
+Cypress.Commands.add('typeUsername', typeUsername);
+
+Cypress.Commands.add('resetUsername', resetUsername);
```
Comment (ojeytonwilliams) [inline cypress/support/commands.ts:63]:
```suggestion
Cypress.Commands.add('resetUsername', resetUsername);

// eslint-disable-next-line @typescript-eslint/no-namespace
declare namespace Cypress {
  interface Chainable {
    login: typeof login;
    preserveSession: typeof preserveSession;
    setPrivacyTogglesToPublic: typeof setPrivacyTogglesToPublic;
    goToSettings: typeof goToSettings;
    typeUsername: typeof typeUsername;
    resetUsername: typeof resetUsername;
  }
}
```
This should do it.

**Review by ojeytonwilliams — APPROVED** [review]
Thanks, @Sembauke!

LGTM :+1:

**Pull request merged**